packages/Search/src/details/DTK_DetailsDistributor.hpp

```diff
@@ -0,0 +1,246 @@
+            "destination_ranks", n );
+        Kokkos::deep_copy( destination_ranks, tmp );
+        _permute = Kokkos::View<int *, Kokkos::HostSpace>( "permute", n );
+        std::vector<int> destinations;
+        extract_and_sort_ranks( destination_ranks, _permute, destinations,
+                                _dest_offsets, _dest_counts );
+
+        if ( _dest_counts.empty() )
+            _dest_offsets.push_back( 0 );
+        else
+            _dest_offsets.push_back( _dest_offsets.back() +
+                                     _dest_counts.back() );
+
+        {
+            int const reorder = 0; // Ranking may *not* be reordered
+            int const sources[1] = {comm_rank};
+            int const degrees[1] = {static_cast<int>( destinations.size() )};
+
+            MPI_Dist_graph_create(
+                _comm, 1, sources, degrees, notNullPtr( destinations.data() ),
+                MPI_UNWEIGHTED, MPI_INFO_NULL, reorder, &_comm_dist_graph );
+        }
+
+        int indegrees;
+        int outdegrees;
+        int weighted;
+        MPI_Dist_graph_neighbors_count( _comm_dist_graph, &indegrees,
+                                        &outdegrees, &weighted );
+        assert( weighted == 0 );
+        assert( outdegrees == (int)destinations.size() );
+
+        std::vector<int> sources( indegrees );
+        MPI_Dist_graph_neighbors( _comm_dist_graph, indegrees,
+                                  notNullPtr( sources.data() ), MPI_UNWEIGHTED,
+                                  outdegrees, notNullPtr( destinations.data() ),
+                                  MPI_UNWEIGHTED );
+
+        _src_counts.resize( indegrees );
+        MPI_Neighbor_alltoall( _dest_counts.data(), 1, MPI_INT,
+                               _src_counts.data(), 1, MPI_INT,
+                               _comm_dist_graph );
+
+        _src_offsets.resize( indegrees + 1 );
+        // exclusive scan
+        _src_offsets[0] = 0;
+        for ( int i = 0; i < indegrees; ++i )
+            _src_offsets[i + 1] = _src_offsets[i] + _src_counts[i];
+
+#if defined( REORDER_RECV )
+        int comm_size = -1;
+        MPI_Comm_size( _comm, &comm_size );
+        _permute_recv.resize( _src_offsets.back() );
+        int offset = 0;
+        for ( int i = 0; i < comm_size; ++i )
+        {
+            auto const it = std::find( sources.begin(), sources.end(), i );
+            if ( it != sources.end() )
+            {
+                int const j = std::distance( sources.begin(), it );
+                std::iota( &_permute_recv[offset],
+                           &_permute_recv[offset] + _src_counts[j],
+                           _src_offsets[j] );
+                offset += _src_counts[j];
+            }
+        }
+        assert( offset == (int)_permute_recv.size() );
+#endif
+
+        return _src_offsets.back();
+    }
+    template <typename View>
+    void doPostsAndWaits( typename View::const_type const &exports,
```

Comment:
Is this the name we want to stick with? I find `doPostsAndWaits` a strange name for a function.

Reply:
I decided to stick with the Tpetra interface for now but intend on revisiting later.  For instance, I don't really like that the class stores two MPI communicators.  I might turn `createFromSends()` into a static member function that builds the `Distributor` which essentially becomes a wrapper for send and receive counts/offsets with a method for executing the communication plan.